### ISSUE TITLE
Timeout while using vscode debugger with any program which spawns a process

### ISSUE THREAD

**Issue by mmohdnowfal** [microsoft/debugpy#1271]
## Environment data

- debugpy version: 1.6.6
- OS and version: Mac ventura 13.3
- Python version : 3.10.9
- Using VS Code or Visual Studio: 
VS Code 
Version: 1.77.2 (Universal)
Commit: e344f1f539a80912a0e9357cec841f36ce97a4e2
Date: 2023-04-06T10:12:31.126Z
Electron: 19.1.11
Chromium: 102.0.5005.196
Node.js: 16.14.2
V8: 10.2.154.26-electron.0
OS: Darwin arm64 22.4.0
Sandboxed: No

## Actual behavior
This issue is similar to what is reported  [here](https://github.com/microsoft/debugpy/issues/1005). 
But this time appears to be the issue is while using vscode debugger with any program which internally uses `os.execv`

We use a library called [Unicon](https://developer.cisco.com/docs/unicon/), for creating a connection to any remote device via ssh.
This library  internally uses `os.execv`, which looks like has a conflict with debugpy(pyadev_monkey.py)

For every new process created using the above library, there is a delay of 5+ seconds and this delays the debugging process and affect the developer productivity and time

## Expected  behavior

Should spawn the process wihout any delay

## Steps to reproduce:
1.  Install the unicon library 
     pip install pyats unicon 
2. Create the below data and python file  
 # Example
  # -------
  #
  #   a simple pyATS testbed YAML file

devices:
    csr1000v-1:
        type: router
        os: iosxe
        credentials:
            default:
                password: foo
                username: foo
            enable:
                password: foo
        connections:
            cli:
                protocol: ssh
                ip: 168.10.1.35
4.  Create a python script 
from pyats.topology import loader

testbed = loader.load('my-testbed.yaml')

device = testbed.devices['csr1000v-1']

device.connect()

device.execute('show version')
5.  Run in debug mode

**Comment by mmohdnowfal** [comment]
Adding the below workaround temporarily disables the issue. Please provide a permanent solution to this 
<img width="1350" alt="Screenshot 2023-04-12 at 3 49 22 PM" src="https://user-images.githubusercontent.com/15832093/231441291-dfd8da0d-63de-43e4-9a6c-394a30464522.png">

**Comment by int19h** [comment]
It's not a conflict; it's an intentional detour of execv to enable debugging of Python subprocesses.

It doesn't seem to be affecting our multiproc tests - is this specific to Unicon?

**Comment by mmohdnowfal** [comment]
> It's not a conflict; it's an intentional detour of execv to enable debugging of Python subprocesses.
> 
> It doesn't seem to be affecting our multiproc tests - is this specific to Unicon?
Yes seems to be,  let me know what to be done here

**Comment by mmohdnowfal** [comment]
Any update on this issue, many of our team members are facing this issue

**Comment by int19h** [comment]
To clarify, do you actually need subprocess debugging in your scenario? If you just want to debug the main process, setting `"subProcess":false` in your debug config will disable all the process creation detouring logic.

**Comment by mmohdnowfal** [comment]
Yes we use subprocess. Ours is an test framework which triggers testcases using subprocesss. 
So using "subprocess":false will not help

**Comment by sstaley-hioscar** [comment]
This has broken my debugging process and yes, I need to be able to debug a subprocess. Will try to work around for now. It would be nice to set my own timeout.

**Comment by bernardo-suez** [comment]
Any news on this? It has been a problem for me too.

**Comment by jose-cubero** [comment]
I am facing the same issue when trying to debug a script that uses `joblib.Parallel`. The work-around `"subprocess":false` does not work for me.

**Comment by dj-kefir-siorbacz** [comment]
yup, impossible to use `joblib.Parallel` in debug mode. I keep seeing this wall of text:
```
61.36s - pydevd: Sending message related to process being replaced timed-out after 5 seconds
62.32s - pydevd: Sending message related to process being replaced timed-out after 5 seconds
63.12s - pydevd: Sending message related to process being replaced timed-out after 5 seconds
64.00s - pydevd: Sending message related to process being replaced timed-out after 5 seconds
65.18s - pydevd: Sending message related to process being replaced timed-out after 5 seconds
66.25s - pydevd: Sending message related to process being replaced timed-out after 5 seconds
67.12s - pydevd: Sending message related to process being replaced timed-out after 5 seconds
68.12s - pydevd: Sending message related to process being replaced timed-out after 5 seconds
69.07s - pydevd: Sending message related to process being replaced timed-out after 5 seconds
70.00s - pydevd: Sending message related to process being replaced timed-out after 5 seconds
71.07s - pydevd: Sending message related to process being replaced timed-out after 5 seconds
71.99s - pydevd: Sending message related to process being replaced timed-out after 5 seconds
72.87s - pydevd: Sending message related to process being replaced timed-out after 5 seconds
73.99s - pydevd: Sending message related to process being replaced timed-out after 5 seconds
75.12s - pydevd: Sending message related to process being replaced timed-out after 5 seconds
76.18s - pydevd: Sending message related to process being replaced timed-out after 5 seconds
77.28s - pydevd: Sending message related to process being replaced timed-out after 5 seconds
78.32s - pydevd: Sending message related to process being replaced timed-out after 5 seconds
79.39s - pydevd: Sending message related to process being replaced timed-out after 5 seconds
80.43s - pydevd: Sending message related to process being replaced timed-out after 5 seconds
81.33s - pydevd: Sending message related to process being replaced timed-out after 5 seconds
82.30s - pydevd: Sending message related to process being replaced timed-out after 5 seconds
83.25s - pydevd: Sending message related to process being replaced timed-out after 5 seconds
84.36s - pydevd: Sending message related to process being replaced timed-out after 5 seconds
85.35s - pydevd: Sending message related to process being replaced timed-out after 5 seconds
86.25s - pydevd: Sending message related to process being replaced timed-out after 5 seconds
87.32s - pydevd: Sending message related to process being replaced timed-out after 5 seconds
88.49s - pydevd: Sending message related to process being replaced timed-out after 5 seconds
89.44s - pydevd: Sending message related to process being replaced timed-out after 5 seconds
90.38s - pydevd: Sending message related to process being replaced timed-out after 5 seconds
91.33s - pydevd: Sending message related to process being replaced timed-out after 5 seconds
92.25s - pydevd: Sending message related to process being replaced timed-out after 5 seconds
93.31s - pydevd: Sending message related to process being replaced timed-out after 5 seconds
94.24s - pydevd: Sending message related to process being replaced timed-out after 5 seconds
95.48s - pydevd: Sending message related to process being replaced timed-out after 5 seconds
96.46s - pydevd: Sending message related to process being replaced timed-out after 5 seconds
97.56s - pydevd: Sending message related to process being replaced timed-out after 5 seconds
98.44s - pydevd: Sending message related to process being replaced timed-out after 5 seconds
99.35s - pydevd: Sending message related to process being replaced timed-out after 5 seconds
```
And tbh it seems like joblib is doing something, but it just so damn slow that it's unusable.

**Comment by Karnabanu** [comment]
Hi @int19h  and @debonte ,

Is there any update on this behaviour?
We are still encountering this issue when using the [Unicon](https://developer.cisco.com/docs/unicon/) library.

Unicon initiates a subprocess to connect to a remote device and waits for the output to appear on the process read buffer.
However, when running with the debugger (debugpy), the issue arises when Unicon creates a subprocess. 
Debugpy detects the subprocess creation at "create_execv" (as mentioned above), but after this detection, debugpy's pydevd prints process replaced message to stdout via the send_process_about_to_be_replaced function.

These additional prints are being captured in the subprocess buffer, causing Unicon to mistakenly interpret them as output from the remote pty session. This results in Unicon's connection entering an inconsistent state.

If there were a CLI option to suppress these extra prints(of pydevd) or redirect them to a file, it could potentially prevent the issue.

**Comment by rchiodo** [comment]
I'm not sure the 'prints' are unnecessary. I believe they are being sent to keep track of the subprocess when it is started.

Here:
https://github.com/microsoft/debugpy/blob/fb0b06cab161478db95420c5bfb485c39d508f62/src/debugpy/adapter/servers.py#L342

The part I don't understand is why this message is going to stdio. It should be on a socket.

**Comment by int19h** [comment]
This is a handler for the DAP `exited` event from pydevd, so it's JSON. The prints, by the looks of it, are coming from pydevd because it cannot deliver the DAP packets for some reason?

**Comment by AndrePostoluck** [comment]
pexpect also spawns subprocesses by its purpose (remote control a process via TTY emulation and output regex scraping) and also sees that 5 second delay.

```
#!/usr/local/bin/python3
import time
import pexpect
t0 = time.perf_counter()
con = pexpect.spawn('zsh')
t1 = time.perf_counter()
print(f"t0-t1: {t1-t0:.6f} seconds")
```

returns: t0-t1: 0.103069 seconds.  
when commenting out send_process_about_to_be_replaced() in new_execv(path, args) in file pydev_monkey.py

returns: t0-t1: 5.156319 seconds
when send_process_about_to_be_replaced() not commented out

**Comment by AndrePostoluck** [comment]
Note that Unicon & pexpect must be doing something different than just a plain subprocess without that problem:
```

#!/usr/local/bin/python3
import time
import subprocess
t0 = time.perf_counter()
result = subprocess.run(["ls", "-l"], capture_output=True, text=True)
t1 = time.perf_counter()
print(f"t0-t1: {t1-t0:.6f} seconds (subprocess)")
```

which returns: t0-t1: 0.004282 seconds (subprocess)
